### DISCUSSION TITLE
inAppNotificationSettings for messaging

### DISCUSSION THREAD

**Discussion by uuouter** [invertase/react-native-firebase#5959] [Q&A] [answered]
Hey all,

First of all, thank you so much for all the hard work on the library and documentation! 👏

Fairly new to React Native, Integrating has been a joy (and granted, a lot of headaches with a lot of hours on StackOverflow / GitHub ;) but mostly my own faults haha).

Apologies if this has been asked/mentioned, couldn't find it.

--

I was wondering how the inAppNotificationSettings for Messaging work / if they are integrated into this library. 
So the option on iOS where you can navigate from the App Settings into your app for Push Notification Settings.

On https://rnfirebase.io/ there is a mention to [Notifee](https://notifee.app/react-native/docs/ios/permissions#in-app-notification-settings) (local notifications) which actually has a section about this:  "This feature is still a work in progress and will be available in a later release." - The library looks cool btw!

As I won't be using local notifications, for now, will Notifee still be required or will this library also receive support for it?

Regards,

Wouter

**Comment by mikehardy** [comment, marked as answer]
Hi there! I'm not sure what it is *exactly* that you are trying to do, so I'm unable to provide any advice really.
In general, react-native-firebase (and firebase as an ecosystem) is built around you as a developer sending JSON via FCM from servers to devices. Firebase handles the sending on the servers, and the firebase library on device handles receipt of FCM JSON, with some very very minimal local-device notification features built in. This is all what we consider "FCM" (cloud messaging) and some people call "push notifications", but I think that makes things less clear because we draw a hard line between the FCM and the actual visible notification a user sees (a local device notification).

But, for anything more than the most basic local device notifications you'll need an actual local device notification package, that's notifee.

To be more specific about your use case you need to detail the JSON you intend to send with FCM, and what you want the device to do (run some logic via a hook? In what application states on what devices? Show some sort of visible local device notification to the user? With what sort of features?)

All the details there become very important, because for instance if you want a custom / branded local device notification to show on iOS when you send an FCM JSON to the device while the app isn't running, you need to use a notification block in the JSON (so iOS will deliver it, it won't deliver data-only JSON reliably to devices when the app is not running) and you need to use Notifee and a notification extension to handle the branding.

If you just want basic notifications a notification block in the FCM JSON may be sufficient. Or perhaps you want data-only for some reason, or mixed payloads. Lots to learn about and specify.

Hope this helps!

**Reply by uuouter** [reply]
Hi Mike,

Thanks for the detailed response! 

I was wondering about the requestPermission({inAppNotificationSettings: true}) for iOS 12+, is it mentioned that:

> When enabled, a button is displayed inside of your applications Settings UI. When pressed, your application is opened and this gives you the chance to navigate to a specific notification settings screen in your UI. This screen could be used to show users how their current notification permissions will impact the usability of the application.

 [image](https://cloud.netlifyusercontent.com/assets/344dbf88-fdf9-42bb-adb4-46f01eedd629/658ab8e1-d0f1-46b2-b092-7f8ed643f4e0/systemnotificationsettings.png) bottom link - [article](https://www.smashingmagazine.com/2018/09/preparing-your-app-for-ios-12-notifications/#deliver-quietly)

Facebook has this, In the notification settings on iOS, there is a link that opens the Facebook app, and shows their own push notification settings.

Currently, I have a screen where users can enable/disable specific notifications. In the iOS Settings menu, I would like to add my link as well, but it is said that this should be done during permission request

**Reply by mikehardy** [reply]
Those settings are in the realm of "local device notifications" because they deal with user presentation of notifications. They don't have anything to do with FCM JSON being delivered to devices, they deal with what happens next. That's how we reason about this split responsibility that Firebase takes on vs local notification packages. Yes, firebase allows you to do request permissions - there is a little gray area, but what you really want is this:

https://notifee.app/react-native/docs/ios/permissions#permission-settings but as you mention it is a work in progress - I've never used it myself so I'm not sure if it is actually working or not? If it is not and if that's vital for your use case we could collaborate in the notifee repo to get it done with you if you proposed a PR. Looks like this is the hook / callback from the system:

```objective-c
func userNotificationCenter(_ center: UNUserNotificationCenter, openSettingsFor notification: UNNotification?) {
    let navController = self.window?.rootViewController as! UINavigationController
    let notificationSettingsVC = NotificationSettingsViewController()
    navController.pushViewController(notificationSettingsVC, animated: true)
}
```

...and somehow we need to expose that to the app, typically that would be a javascript callback like `onNotificationSettingsOpened` or something, then you would do a javascript app-level implementation that set a handler in to that javascript callback, and when it executed you would do something like 

```javascript
const onNotificationSettingsOpenedHandler = () => {
  const navigator = useNavigation();
  navigator.navigate("NotificationSettingsScreen");

  // or maybe, for react-navigation web-compatible linking / deep-linking:
  const linkTo = useLinking();
  linkTo('/settings/notifications');
}
```

(just coding that up in a text area, no idea if it's right on)

I also use react-native-permissions for permission management, and I note they don't seem to support this in-app settings screen either: https://github.com/zoontek/react-native-permissions#requestnotifications

**Reply by uuouter** [reply]
Hey Mike,

Thanks again for your response and investigation (also for checking react-native-permissions).

> That's how we reason about this split responsibility that Firebase takes on vs local notification packages. Yes, firebase allows you to do request permissions - there is a little gray area, but what you really want is this:

That split makes perfectly sense!

Taking a look at Notifee today, I've got some free time coming up after the holidays and would love to collaborate on it. The feature seems small and do-able, probably will need some pointers on best practices (the code above already helps!), but we can do that in the PR at the Notifee repo :)

Happy holidays!

Wouter

**Reply by mikehardy** [reply]
Hey @helenaford - nothing urgent, but I did a quick (in text area coding/design ⚠️ 😅 ) sketch on how to support a new iOS request permissions feature / system callback hook in Notifee, and if you have any general thoughts on it with an eye towards real concrete specifics on implementation, that might make a PR from @uuouter achievable and get a cool feature in Notifee)
